actors/builtin/miner/miner_actor.go

```diff
@@ -210,42 +220,51 @@ func (a Actor) ChangeMultiaddrs(rt Runtime, params *ChangeMultiaddrsParams) *adt
 // WindowedPoSt //
 //////////////////
 
+type PoStPartition struct {
+	// Partitions are numbered per-deadline, from zero.
+	Index uint64
+	// Sectors skipped while proving that weren't already declared faulty
+	Skipped *abi.BitField
+}
+
 // Information submitted by a miner to provide a Window PoSt.
 type SubmitWindowedPoStParams struct {
 	// The deadline index which the submission targets.
 	Deadline uint64
-	// The partition indices being proven.
-	// Partitions are counted across all deadlines, such that all partition indices in the second deadline are greater
-	// than the partition numbers in the first deadlines.
-	Partitions []uint64
+	// The partitions being proven.
+	Partitions []PoStPartition
 	// Array of proofs, one per distinct registered proof type present in the sectors being proven.
 	// In the usual case of a single proof type, this array will always have a single element (independent of number of partitions).
 	Proofs []abi.PoStProof
-	// Sectors skipped while proving that weren't already declared faulty
-	Skipped abi.BitField
 }
 
 // Invoked by miner's worker address to submit their fallback post
 func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams) *adt.EmptyValue {
 	currEpoch := rt.CurrEpoch()
 	store := adt.AsStore(rt)
 	var st State
-	var newFaultSectors []*SectorOnChainInfo
-	var recoveredSectors []*SectorOnChainInfo
-	penalty := abi.NewTokenAmount(0)
+
+	if params.Deadline > WPoStPeriodDeadlines {
+		rt.Abortf(exitcode.ErrIllegalArgument, "invalid deadline %d of %d", params.Deadline, WPoStPeriodDeadlines)
+	}
+	// TODO: limit the length of proofs array https://github.com/filecoin-project/specs-actors/issues/416
```

Comment:
@anorth, this is done, right?

https://github.com/filecoin-project/specs-actors/pull/648/files#diff-e47fb4c42d791672c9550edb1dc5f969R267

Reply:
We didn't make it worse, but I think still something to do here.